File: go/hot_100/3_longest_substring_without_repeating_characters.go
package main

func Max(a int, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}

func lengthOfLongestSubstring(s string) int {
	if len(s) == 0 {
		return 0
	}
	a, b, res := 0, 0, 1
	count := make(map[byte]int)
	count[s[b]] = 0
	for b < len(s)-1 {
		b++
		v, ok := count[s[b]]
		if ok {
			for a <= v {
				delete(count, s[a])
				a++
			}
		}
		count[s[b]] = b
		res = Max(res, b-a+1)
	}
	return res
}

// func main() {
// 	lengthOfLongestSubstring("abcabcbb")
// }

File: python/hot_100/3_longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s)<=1:
            return len(s)
        l=0
        r=1
        d=dict()
        res=1
        d[s[l]]=0
        while r<len(s):
            while r<len(s) and s[r] not in d.keys():
                d[s[r]]=r
                r+=1
            res=max(res,r-d[s[l]])
            while l<r and r<len(s) and s[l]!=s[r]:
                del d[s[l]]
                l+=1
            del d[s[l]]
            l+=1
        return res
        
s=Solution()
print(s.lengthOfLongestSubstring("abcabcbb"))